cli/standalone/virtual_fs.rs

```diff
@@ -0,0 +1,983 @@
+  }
+
+  fn find_entry_inner<'a>(
+    &'a self,
+    path: &Path,
+    seen: &mut HashSet<PathBuf>,
+  ) -> std::io::Result<(PathBuf, VfsEntryRef<'a>)> {
+    let mut path = Cow::Borrowed(path);
+    loop {
+      let (resolved_path, entry) =
+        self.find_entry_no_follow_inner(&path, seen)?;
+      match entry {
+        VfsEntryRef::Symlink(symlink) => {
+          if !seen.insert(path.to_path_buf()) {
+            return Err(std::io::Error::new(
+              std::io::ErrorKind::Other,
+              "circular symlinks",
+            ));
+          }
+          path = Cow::Owned(symlink.resolve_dest_from_root(&self.root_path));
+        }
+        _ => {
+          return Ok((resolved_path, entry));
+        }
+      }
+    }
+  }
+
+  fn find_entry_no_follow(
+    &self,
+    path: &Path,
+  ) -> std::io::Result<(PathBuf, VfsEntryRef)> {
+    self.find_entry_no_follow_inner(path, &mut HashSet::new())
+  }
+
+  fn find_entry_no_follow_inner<'a>(
+    &'a self,
+    path: &Path,
+    seen: &mut HashSet<PathBuf>,
+  ) -> std::io::Result<(PathBuf, VfsEntryRef<'a>)> {
+    let relative_path = match path.strip_prefix(&self.root_path) {
+      Ok(p) => p,
+      Err(_) => {
+        return Err(std::io::Error::new(
+          std::io::ErrorKind::NotFound,
+          "path not found",
+        ));
+      }
+    };
+    let mut final_path = self.root_path.clone();
+    let mut current_entry = VfsEntryRef::Dir(&self.dir);
+    for component in relative_path.components() {
+      let component = component.as_os_str().to_string_lossy();
+      let current_dir = match current_entry {
+        VfsEntryRef::Dir(dir) => {
+          final_path.push(component.as_ref());
+          dir
+        }
+        VfsEntryRef::Symlink(symlink) => {
+          let dest = symlink.resolve_dest_from_root(&self.root_path);
+          let (resolved_path, entry) = self.find_entry_inner(&dest, seen)?;
+          final_path = resolved_path; // overwrite with the new resolved path
+          match entry {
+            VfsEntryRef::Dir(dir) => {
+              final_path.push(component.as_ref());
+              dir
+            }
+            _ => {
+              return Err(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                "path not found",
+              ));
+            }
+          }
+        }
+        _ => {
+          return Err(std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            "path not found",
+          ));
+        }
+      };
+      match current_dir
+        .entries
+        .binary_search_by(|e| e.name().cmp(&component))
+      {
+        Ok(index) => {
+          current_entry = current_dir.entries[index].as_ref();
+        }
+        Err(_) => {
+          return Err(std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            "path not found",
+          ));
+        }
+      }
+    }
+
+    Ok((final_path, current_entry))
+  }
+}
+
+#[derive(Clone)]
+struct FileBackedVfsFile {
```

Comment:
Something we may want to consider is storing the files compressed, then decompressing them on open.

Reply:
Sounds like a good idea, but we should benchmark it - might not be worth added complexity and startup time 👍